src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs

```diff
@@ -1155,24 +1177,24 @@ arraysOf n (D.Stream step state) =
             error $ "Streamly.Internal.Data.Array.Foreign.Mut.Type.arraysOf: "
                     ++ "the size of arrays [" ++ show n
                     ++ "] must be a natural number"
-        Array start end bound <- liftIO $ newArray n
-        return $ D.Skip (GroupBuffer st start end bound)
+        Array contents start end bound <- liftIO $ newArray n
+        return $ D.Skip (GroupBuffer st contents start end bound)
 
-    step' gst (GroupBuffer st start end bound) = do
+    step' gst (GroupBuffer st contents start end bound) = do
         r <- step (adaptState gst) st
         case r of
             D.Yield x s -> do
                 liftIO $ poke end x
                 let end' = end `plusPtr` sizeOf (undefined :: a)
                 return $
                     if end' >= bound
-                    then D.Skip (GroupYield start end' bound (GroupStart s))
-                    else D.Skip (GroupBuffer s start end' bound)
-            D.Skip s -> return $ D.Skip (GroupBuffer s start end bound)
-            D.Stop -> return $ D.Skip (GroupYield start end bound GroupFinish)
+                    then D.Skip (GroupYield contents start end' bound (GroupStart s))
+                    else D.Skip (GroupBuffer s contents start end' bound)
+            D.Skip s -> return $ D.Skip (GroupBuffer s contents start end bound)
+            D.Stop -> return $ D.Skip (GroupYield contents start end bound GroupFinish)
 
-    step' _ (GroupYield start end bound next) =
-        return $ D.Yield (Array start end bound) next
+    step' _ (GroupYield contents start end bound next) =
+        return $ D.Yield (Array contents start end bound) next
```

Comment:
Don't we ever touch foreign ptr ever here? (Question for the original code)

Reply:
Looks like a bug. But may not occur in practice depending on how the array is used. We can try adding a touch and check if it degrades perf.